nxcomp/Log.cpp

```diff
@@ -0,0 +1,121 @@
+/**************************************************************************/
+/*                                                                        */
+/* Copyright (c) 2001, 2011 NoMachine (http://www.nomachine.com)          */
+/* Copyright (c) 2008-2014 Oleksandr Shneyder <o.shneyder@phoca-gmbh.de>  */
+/* Copyright (c) 2014-2016 Ulrich Sibiller <uli42@gmx.de>                 */
+/* Copyright (c) 2014-2016 Mihai Moldovan <ionic@ionic.de>                */
+/* Copyright (c) 2011-2016 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>*/
+/* Copyright (c) 2015-2016 Qindel Group (http://www.qindel.com)           */
+/*                                                                        */
+/* NXCOMP, NX protocol compression and NX extensions to this software     */
+/* are copyright of the aforementioned persons and companies.             */
+/*                                                                        */
+/* Redistribution and use of the present software is allowed according    */
+/* to terms specified in the file LICENSE.nxcomp which comes in the       */
+/* source distribution.                                                   */
+/*                                                                        */
+/* All rights reserved.                                                   */
+/*                                                                        */
+/* NOTE: This software has received contributions from various other      */
+/* contributors, only the core maintainers and supporters are listed as   */
+/* copyright holders. Please contact us, if you feel you should be listed */
+/* as copyright holder, as well.                                          */
+/*                                                                        */
+/**************************************************************************/
+
+
+#include <stdio.h>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <iomanip>
+
+#include "Log.h"
+
+NXLog nx_log;
+
+
+bool NXLog::will_log() const
+{
+	std::map<std::string, NXLogLevel>::const_iterator item = per_file_levels_.find(current_file());
+
+	if ( item != per_file_levels_.end() )
+	{
+		return current_level() >= item->second;
+	}
+	else
+	{
+		return current_level() >= level();
+	}
+}
+
+
+std::string NXLog::stamp_to_string(const NXLogStamp& stamp) const
+{
+	std::ostringstream oss;
+
+	static const char* level_names[] = {
+		"DEBUG",
+		"INFO ",
+		"WARN ",
+		"ERROR",
+		"FATAL"
+	};
+
+	if ( log_level() )
+		oss << ((stamp.level() >=0 && stamp.level() < NXLOG_LEVEL_COUNT ) ? level_names[stamp.level()] : "???") << " ";
+
+	if ( log_time() )
+	{
+		struct timeval timestamp = stamp.timestamp();
+		struct tm timeinfo;
+
+		localtime_r(&timestamp.tv_sec, &timeinfo);
+
+		if ( log_unix_time() )
+		{
+			oss << timestamp.tv_sec;
+		}
+		else
+		{
+			#if __cplusplus >= 201103L && (!defined(__GNUC__) || __GNUC__ >= 5)
```

Comment:
Hmm... yeah, but `clang` also defined __GNUC__ and sets it to a rather old value (4, I guess?)

There are lot of compilers out there, sadly, but... let's try something like that instead:

```C++
  #if __cplusplus >= 201103L && ((defined(_MSC_VER) && (_MSC_VER >= 1700)) || (defined(__clang__)) || (defined(__GNUC__) && (__GNUC__ >= 5)) || (!defined(_MSC_VER) && (!defined(__clang__) && (!defined(__GNUC__))))
```

Hopefully I didn't miss anything? Like I said, there are lot of other compilers, the last block is a catch-all. Maybe this could be optimized, but I didn't have a better idea.